Fix crash when accessing unbound textures in materials

kh-huawei opened the pull request:
## What does this PR do?

Potential fix for [PR 19208](https://github.com/o3de/o3de/issues/19208)

When a material accesses a texture that is not bound, it currently accesses whatever texture is bound at bindless read index 0.
If there is nothing bound, vulkan systems are likely to crash.

This PR explicitly binds a Null - texture and propagates the bindless read index to the materials.

Note that this required a manual re-processing of all material shaders, since changing the MaterialSRG - files did not trigger any rebuilds automatically.

## How was this PR tested?

Vulkan / DX12 on Windows only. I was not able to reproduce the linked issue, since that occurs only on linux.

kh-huawei committed:
Fix crash when accessing unbound textures in materials

Signed-off-by: Karl Haubenwallner <karl.haubenwallner@huawei.com>

nick-l-o3de commented:
I was able to verify that this does fix the emissive crash.

nick-l-o3de commented:
```
[2025-08-28T16:29:46.574Z] /data/workspace/o3de/Gems/Atom/RPI/Code/Source/RPI.Public/Material/MaterialSystem.cpp:710:34: error: no member named 'ImageSystemInterface' in namespace 'AZ::RPI'; did you mean 'BufferSystemInterface'?
[2025-08-28T16:29:46.574Z]             auto pool = AZ::RPI::ImageSystemInterface::Get()->GetSystemAttachmentPool();
[2025-08-28T16:29:46.574Z]                         ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
[2025-08-28T16:29:46.574Z]                                  BufferSystemInterface
[2025-08-28T16:29:46.574Z] /data/workspace/o3de/Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Buffer/BufferSystemInterface.h:66:35: note: 'BufferSystemInterface' declared here
[2025-08-28T16:29:46.574Z]         class ATOM_RPI_PUBLIC_API BufferSystemInterface
[2025-08-28T16:29:46.574Z]                                   ^
[2025-08-28T16:29:46.574Z] /data/workspace/o3de/Gems/Atom/RPI/Code/Source/RPI.Public/Material/MaterialSystem.cpp:710:63: error: no member named 'GetSystemAttachmentPool' in 'AZ::RPI::BufferSystemInterface'
[2025-08-28T16:29:46.574Z]             auto pool = AZ::RPI::ImageSystemInterface::Get()->GetSystemAttachmentPool();
[2025-08-28T16:29:46.574Z]                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^
```

Maybe a missing header?

Also, weirdly enough, the shader kills dxc.exe on android and windows, without explaining why.

I was testing on linux



```
Info | origin: AssetBuilder | ShaderPlatformInterface: AZ::RHI::ExecuteShaderCompiler elapsedTimeMillis: 89
Info | origin: AssetBuilder | ShaderAssetBuilder: Note: This shader uses 1 of 128 available shader variant key bits.
Info | origin: AssetBuilder | ShaderVariantAssetBuilder: Variant StableId: 0
Info | origin: AssetBuilder | ShaderVariantAssetBuilder: Variant Shader Options: DefaultOption=?
Info | origin: AssetBuilder | ShaderVariantAssetBuilder: Entry Point: MainVS
Info | origin: AssetBuilder | ShaderVariantAssetBuilder: Begin compiling shader function "MainVS"
Info | origin: AssetBuilder | ShaderPlatformInterface: Executing '"D:\workspace\o3de\build\windows\bin\profile\Builders\DirectXShaderCompiler\dxc.exe" -E MainVS -T vs_6_2 -Zpr -enable-16bit-types -Fo "D:\workspace\o3de\AutomatedTesting\user\AssetProcessorTemp\JobTemp-nJCshZ\SceneMaterialSrg_dx12.dxil.bin" -Fh "D:\workspace\o3de\AutomatedTesting\user\AssetProcessorTemp\JobTemp-nJCshZ\SceneMaterialSrg_dx12.dxil.txt" "D:\workspace\o3de\AutomatedTesting\user\AssetProcessorTemp\JobTemp-nJCshZ\SceneMaterialSrg_dx12.hlsl.prepend"' ...
ERROR | origin: AssetBuilder | DXC: dxc failed : error code 0x8007045a.
ERROR | origin: AssetBuilder | ShaderPlatformInterface: DXC has exited with error code 1
```

error code 0x8007045a indicates a DLL missing issue, maybe somewthing went wrong with updated DXC on jenkins?

kh-huawei committed:
added missing include

Signed-off-by: Karl Haubenwallner <karl.haubenwallner@huawei.com>

kh-huawei commented:
I've added the missing include for the non-unity build.
About the dxc - error: no idea, it works locally, and the shader isn't complicated enough to cause this - maybe the windows version of the CI system is off? 
The error code on windows is the same as on linux: `DXC: dxc failed : error code 0x8007045a.`
```
[2025-08-28T16:31:42.227Z] AssetProcessor: JOB LOG: 8/28/2025 4:32 PM | Info | origin: AssetBuilder | ShaderPlatformInterface: Executing '"D:\workspace\o3de\build\windows\bin\profile\Builders\DirectXShaderCompiler\dxc.exe" -E MainVS -T vs_6_2 -Zpr -enable-16bit-types -Fo "D:\workspace\o3de\AutomatedTesting\user\AssetProcessorTemp\JobTemp-MQMHtO\SceneMaterialSrg_dx12.dxil.bin" -Fh "D:\workspace\o3de\AutomatedTesting\user\AssetProcessorTemp\JobTemp-MQMHtO\SceneMaterialSrg_dx12.dxil.txt" "D:\workspace\o3de\AutomatedTesting\user\AssetProcessorTemp\JobTemp-MQMHtO\SceneMaterialSrg_dx12.hlsl.prepend"' ...
[2025-08-28T16:31:42.227Z] AssetProcessor: JOB LOG: 8/28/2025 4:32 PM | Context | origin: AssetBuilder | C: [Platform API] = dx12
[2025-08-28T16:31:42.227Z] AssetProcessor: JOB LOG: 8/28/2025 4:32 PM | Context | origin: AssetBuilder | C: [Trace] = D:/workspace/o3de/Gems/Atom/RHI/Code/Source/RHI.Edit/Utils.cpp(395): 'bool __cdecl AZ::RHI::ReportMessages(class AZStd::basic_string_view<char,struct AZStd::char_traits<char> >,class AZStd::basic_string_view<char,struct AZStd::char_traits<char> >,bool)'
[2025-08-28T16:31:42.227Z] AssetProcessor: JOB LOG: 8/28/2025 4:32 PM | Context | origin: AssetBuilder | C: [Type] = Trace::Error
[2025-08-28T16:31:42.227Z] AssetProcessor: JOB LOG: 8/28/2025 4:32 PM | ERROR | origin: AssetBuilder | DXC: dxc failed : error code 0x8007045a.
```

nick-l-o3de commented:
> I've added the missing include for the non-unity build. About the dxc - error: no idea, it works locally, and the shader isn't complicated enough to cause this - maybe the windows version of the CI system is off? The error code on windows is the same as on linux: `DXC: dxc failed : error code 0x8007045a.`
> 
> ```
> [2025-08-28T16:31:42.227Z] AssetProcessor: JOB LOG: 8/28/2025 4:32 PM | Info | origin: AssetBuilder | ShaderPlatformInterface: Executing '"D:\workspace\o3de\build\windows\bin\profile\Builders\DirectXShaderCompiler\dxc.exe" -E MainVS -T vs_6_2 -Zpr -enable-16bit-types -Fo "D:\workspace\o3de\AutomatedTesting\user\AssetProcessorTemp\JobTemp-MQMHtO\SceneMaterialSrg_dx12.dxil.bin" -Fh "D:\workspace\o3de\AutomatedTesting\user\AssetProcessorTemp\JobTemp-MQMHtO\SceneMaterialSrg_dx12.dxil.txt" "D:\workspace\o3de\AutomatedTesting\user\AssetProcessorTemp\JobTemp-MQMHtO\SceneMaterialSrg_dx12.hlsl.prepend"' ...
> [2025-08-28T16:31:42.227Z] AssetProcessor: JOB LOG: 8/28/2025 4:32 PM | Context | origin: AssetBuilder | C: [Platform API] = dx12
> [2025-08-28T16:31:42.227Z] AssetProcessor: JOB LOG: 8/28/2025 4:32 PM | Context | origin: AssetBuilder | C: [Trace] = D:/workspace/o3de/Gems/Atom/RHI/Code/Source/RHI.Edit/Utils.cpp(395): 'bool __cdecl AZ::RHI::ReportMessages(class AZStd::basic_string_view<char,struct AZStd::char_traits<char> >,class AZStd::basic_string_view<char,struct AZStd::char_traits<char> >,bool)'
> [2025-08-28T16:31:42.227Z] AssetProcessor: JOB LOG: 8/28/2025 4:32 PM | Context | origin: AssetBuilder | C: [Type] = Trace::Error
> [2025-08-28T16:31:42.227Z] AssetProcessor: JOB LOG: 8/28/2025 4:32 PM | ERROR | origin: AssetBuilder | DXC: dxc failed : error code 0x8007045a.
> ```

Yeah its weird that it would die on both platforms in the same way.  I'll try building with a clean cache.
Didn't someone rev our DXC plugin recently though?

nick-l-o3de commented:
I think the DXC plugin was renewed recently and the jenkins machines have issues with it.  The windows GHA machines are less problematic

nick-l-o3de commented:
We're currently working on fixing the DXC issue, its that the jenkins machine is on ubuntu 20.04 and the dxc compiler was built for 22.04 and thus the GLIBC is incompatible.  (We're reving our jenkins to 22.04 if we can.).

Same issue on windows - the DXC was built on vs2022 vc latest, and jenkins runs the oldest supported vs2019 to make sure that we don't publish something incompatible.  In this case, it was incompat.  A new DXC is ready to go and solves the windows isuse, but the linux issue will require infra updates to the jenkins nodes, which is still forthcoming.  ( we are not going to downgrade DXC's compile all the way down to ubuntu 20.04, that is out of support now)

nick-l-o3de commented:
We have sorted out the various issues causing the AR/CI build to fail (unrelated to this), Please rebase onto latest `development` or merge it in, and we can try to get this pushed in.

nick-l-o3de commented:
This is causing a crash in AR, because the tests don't call RPISystem::Initialize (which initializes the material system), instead, they call `RPISystem::InitializeForTests` which does not initialize the material system.   Previously I guess it was not necessary to initialize the material system, but now it is (?) since tests seem to end up in it.

Additionally, the material system does this in its Init() function:
```cppp
 m_nullTexture = AZ::RPI::AttachmentImage::Create(
     *pool.get(), nullImageDesc, AZ_NAME_LITERAL("MaterialNullTexture"), &nullImageClearValue, &imageViewDesc);
     ```
but it does not clear this instance during Shutdown() causing an additional "you have leaked resources!" warnings.

Consider adding the call to initialize the material system in `RPISystem::InitializeForTests`
If you do, please add 
```cpp
 if (m_nullTexture)
 {
     m_nullTexture.reset();
 }
```
in MaterialSystem::Shutdown

kh-huawei committed:
fix testcases

Signed-off-by: Karl Haubenwallner <karl.haubenwallner@huawei.com>

kh-huawei commented:
Initialization of the material-system previously happened with the RPI - system, but with this PR i needed to move it after the ImageSystem, since it is now creating an empty texture.

I've rebased the PR and fixed the testcases now.
